### ISSUE TITLE
Issue with Exporting Interpolate to ONNX 

### ISSUE THREAD

**Issue by lara-hdr** [pytorch/pytorch#27176]
torch.functional.interpolate allows users to choose between scale_factors and output_size.

In case scale_factors is provided, the output_size is computed in interpolate() in torch/nn/functional.py and will be used from this point, since the aten operators aten::upsample_[mode][dim]d only provide the output_size.
Like the aten ops, the ONNX symbolic functions also provide the output_size (and not scale_factors).

The issue is that ONNX Resize/Upsample requires the scales rather than the output_size (until opset 11). 
So when an interpolate (with provided scale_factors) is being exported, the scale_factors are being converted to output_size (in interpolate() in functional.py) and then the computed output_size is being converted back to scale_factors (in the onnx symbolic), which creates a subgraph of multiple nodes that can be avoided.

This last operation (output_size -> scale_factors) requires a float division which results in floating-point errors and a loss in the precision of the computed scales in some cases.
We are seeing such problems in the Fast Neural Style Transfer model in https://github.com/onnx/models/tree/master/vision/style_transfer/fast_neural_style.

We could fix this issue if we can access the inputted scale_factors when exporting interpolate.
For that, we will need to modify the aten ops 'aten::upsample_[mode][dim]d' to provide both output_size and scale_factors, and directly use the scale_factors in the export when provided.

Would there be any issues with modifying the upsample aten operators by adding scale_factors as an additional argument?

cc @yf225

**Comment by houseroad** [comment]
cc: @gchanan for aten api change

**Comment by lara-hdr** [comment]
I noticed that when scripting interpolate, aten::__interpolate will be used rather than aten::upsample_[mode][dim]d.
And aten::__interpolate provides both the output_size and the scale_factor and could solve our issue.

Why the distinction between the two? and why isn't aten::__interpolate used in tracing as well?

**Comment by lara-hdr** [comment]
@gchanan, @houseroad, any thoughts on this?

**Comment by lara-hdr** [comment]
Hey @gchanan, this is the issue that we discussed regarding interpolate.
Thanks!

**Comment by lara-hdr** [comment]
@gchanan any thoughts?

**Comment by gchanan** [comment]
@lara-hdr: sorry for the slow response, getting caught back up after the release and DevCon.

It generally seems okay to take both, if for no other reason than to match up the C++ and the python API.

There is an issue with serialized torchscript models though, and it's not obvious to me how this will be done without an adapter. Maybe @suo or @houseroad can provide some guidance on how to do this.

**Comment by glenn-jocher** [comment]
I can not export from PyTorch to ONNX using the upsample operator with torch version 1.3.1, 
onnx version 1.5.0.

MINIMUM CODE TO REPRODUCE:
```python
import torch
import torch.nn as nn
import torch.nn.functional as F
import onnx

print(torch.__version__)
print(onnx.__version__)


class TestModel(nn.Module):
    def __init__(self):
        super(TestModel, self).__init__()

    def forward(self, x):
        x = F.interpolate(x, scale_factor=2, mode='nearest')
        return x


torch_model = TestModel()
dummy_input = torch.randn(1, 3, 256, 256)

torch_out = torch.onnx.export(torch_model, dummy_input, 'model.onnx', verbose=True, opset_version=11)

onnx_model = onnx.load('model.onnx')
print(onnx_model)
onnx.checker.check_model(onnx_model)
```

Produces the following error:
```bash
Traceback (most recent call last):
  File "/Users/glennjocher/PycharmProjects/iD/upsample_error_reproduce.py", line 26, in <module>
    onnx.checker.check_model(onnx_model)
  File "/Users/glennjocher/.conda/envs/yolov3/lib/python3.7/site-packages/onnx/checker.py", line 86, in check_model
    C.check_model(model.SerializeToString())
onnx.onnx_cpp2py_export.checker.ValidationError: Node () has input size 4 not in range [min=2, max=2].

==> Context: Bad node spec: input: "input" input: "23" input: "23" input: "22" output: "24" op_type: "Resize" attribute { name: "coordinate_transformation_mode" s: "asymmetric" type: STRING } attribute { name: "cubic_coeff_a" f: -0.75 type: FLOAT } attribute { name: "mode" s: "nearest" type: STRING } attribute { name: "nearest_mode" s: "floor" type: STRING }
```

**Comment by blueardour** [comment]
Hi, I successfully convert the upsample with the scale param, however, I failed to pass the result test.

I refer https://github.com/pytorch/pytorch/issues/18113 for a simple test code:

`import torch.nn as nn
import torch
import torch.nn.functional as F
import numpy as np

import onnx
import onnxruntime
import caffe2.python.onnx.backend as backend

class Test(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = nn.Conv2d(3, 64, 3)

    def forward(self, x):
        x = F.interpolate(x, scale_factor=2, mode='bilinear', align_corners=False) #no warning, all clear
        #x = F.interpolate(x, size=(400, 600), mode='bilinear', align_corners=False) #no warning, all clear
        #x = self.conv1(x)
        return x

model = Test()
model.eval()
x = torch.rand((1, 3, 200, 300))
torch.onnx.export(model, x, "test.onnx", verbose=True,keep_initializers_as_inputs=True)

with torch.no_grad():
    torch_out = model(x)

ort_session = onnxruntime.InferenceSession("test.onnx")
ort_input = {ort_session.get_inputs()[0].name: x.cpu().numpy()}
ort_out = ort_session.run(None, ort_input)[0]
try:
    np.testing.assert_allclose(torch_out.cpu().numpy(), ort_out, rtol=1e-03, atol=1e-05)
except AssertionError as e:
    print(e)

load_model = onnx.load("test.onnx")
onnx_model = backend.prepare(load_model)
outputs = onnx_model.run({load_model.graph.input[0].name: x.data.numpy()})
caffe2 = outputs[0]
try:
    np.testing.assert_allclose(torch_out.cpu().numpy(), caffe2, rtol=1e-03, atol=1e-05)
except AssertionError as e:
    print(e)`

**Comment by blueardour** [comment]
Hi, I successfully convert the mode, however didn't pass the result test. Someone also report the issue.

https://github.com/pytorch/pytorch/issues/18113#issuecomment-608318113

Is there any advise on testing?

**Comment by yashnv** [comment]
Hello @lara-hdr, is this still an issue? The scale parameter still outputs as None and goes to the output-size scope for the default version of `upsample_nearest2d`.
As a workaround for exporting `torch.nn.functional.interpolate(x, scale_factor=2)`, I am using this before calling `torch.onnx.export`:

For opset11:
```
import torch.onnx.symbolic_opset11 as onnx_symbolic
def upsample_nearest2d(g, input, output_size, *args):
    # Currently, TRT 7.0/7.1 ONNX Parser does not support all ONNX ops
    # needed to support dynamic upsampling ONNX forumlation
    # Here we hardcode scale=2 as a temporary workaround
    scales = g.op("Constant", value_t=torch.tensor([1., 1., 2., 2.]))
    empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
    return g.op("Resize", input, empty_tensor, scales, mode_s="nearest", nearest_mode_s="floor")

onnx_symbolic.upsample_nearest2d = upsample_nearest2d
```
For opset10:
```
import torch.onnx.symbolic_opset10 as onnx_symbolic
def upsample_nearest2d(g, input, output_size, *args):
    # Currently, TRT 5.1/6.0/7.0 ONNX Parser does not support all ONNX ops
    # needed to support dynamic upsampling ONNX forumlation
    # Here we hardcode scale=2 as a temporary workaround
    scales = g.op("Constant", value_t=torch.tensor([1., 1., 2., 2.]))
    return g.op("Resize", input, scales, mode_s="nearest")

onnx_symbolic.upsample_nearest2d = upsample_nearest2d
```

**Comment by shi-yan** [comment]
I'm unable to export interpolate 

nn.functional.interpolate(
                pool_scale(conv5),
                scale_factor=scale_factor,
                mode='bilinear', align_corners=False, recompute_scale_factor=False))

```
  File "test.py", line 232, in <module>
    main(cfg, args.gpu)
  File "test.py", line 133, in main
    torch.onnx.export(segmentation_module_dump, dummy_input, "segmentation.onnx", verbose=True, export_params=True,  opset_version=12, do_constant_folding=True,  input_names=input_names, output_names=output_names )
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/__init__.py", line 271, in export
    return utils.export(model, args, f, export_params, verbose, training,
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/utils.py", line 88, in export
    _export(model, args, f, export_params, verbose, training, input_names, output_names,
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/utils.py", line 694, in _export
    _model_to_graph(model, args, verbose, input_names,
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/utils.py", line 463, in _model_to_graph
    graph = _optimize_graph(graph, operator_export_type,
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/utils.py", line 206, in _optimize_graph
    graph = torch._C._jit_pass_onnx(graph, operator_export_type)
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/__init__.py", line 309, in _run_symbolic_function
    return utils._run_symbolic_function(*args, **kwargs)
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/utils.py", line 997, in _run_symbolic_function
    return symbolic_fn(g, *inputs, **attrs)
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/symbolic_opset9.py", line 930, in symbolic_fn
    return sym_help._onnx_unsupported(name + ', since output size is not factor of input size')
  File "/home/shiy/.local/lib/python3.8/site-packages/torch/onnx/symbolic_helper.py", line 224, in _onnx_unsupported
    raise RuntimeError('Unsupported: ONNX export of operator {}. '
RuntimeError: Unsupported: ONNX export of operator adaptive_avg_pool2d, since output size is not factor of input size. Please feel free to request support or submit a pull request on PyTorch GitHub.
shiy@19e40db3aa4b:~/adobe_test/semantic-segmentation-pytorch$ pip list

```

any solutions to this issue?

**Comment by garymm** [comment]
Supposedly fixed by #28324. Please open a new issue with detailed repro instructions if there's still a problem.

**Comment by zshn25** [comment]
@yashnv I tried your approach
> ```
> import torch.onnx.symbolic_opset11 as onnx_symbolic
>def upsample_nearest2d(g, input, output_size, *args):
>   # Currently, TRT 7.0/7.1 ONNX Parser does not support all ONNX ops
>   # needed to support dynamic upsampling ONNX forumlation
>    # Here we hardcode scale=2 as a temporary workaround
>    scales = g.op("Constant", value_t=torch.tensor([1., 1., 2., 2.]))
>    empty_tensor = g.op("Constant", value_t=torch.tensor([], dtype=torch.float32))
>    return g.op("Resize", input, empty_tensor, scales, mode_s="nearest", nearest_mode_s="floor")
>
>onnx_symbolic.upsample_nearest2d = upsample_nearest2d
>```
 but I get 

```bash
ValueError                                Traceback (most recent call last)
Cell In[8], line 26
     24 model = QuantizedDepth(encoder=encoder, decoder=depth_decoder, for_quantization=False).eval().to(device)
     25 scripted_model = torch.jit.script(model, input_image_pytorch.to(device).detach())
---> 26 torch.onnx.export(scripted_model.to(device),               # model being run
     27                   input_image_pytorch.to(device),                         # model input (or a tuple for multiple inputs)
     28                   "scripted_monodepth2.onnx",   # where to save the model (can be a file or file-like object)
     29                   export_params=True,        # store the trained parameter weights inside the model file
     30                   opset_version=11,          # the ONNX version to export the model to
     31                   # do_constant_folding=True,  # whether to execute constant folding for optimization. buggy in torch == 1.12
     32                   input_names = ['input'],   # the model's input names
     33                   output_names = ['output'], # the model's output names
     34                 #   dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
     35                 #                 'output' : {0 : 'batch_size'}}
     36                                 )
     37 print("Size (MB):", os.path.getsize("traced_monodepth2.onnx")/1e6)
     38 print("Size (MB):", os.path.getsize("monodepth2.onnx")/1e6)

File ~/miniconda3/envs/inference/lib/python3.8/site-packages/torch/onnx/utils.py:504, in export(model, args, f, export_params, verbose, training, input_names, output_names, operator_export_type, opset_version, do_constant_folding, dynamic_axes, keep_initializers_as_inputs, custom_opsets, export_modules_as_functions)
    186 @_beartype.beartype
    187 def export(
    188     model: Union[torch.nn.Module, torch.jit.ScriptModule, torch.jit.ScriptFunction],
   (...)
    204     export_modules_as_functions: Union[bool, Collection[Type[torch.nn.Module]]] = False,
    205 ) -> None:
    206     r"""Exports a model into ONNX format.
    207 
    208     If ``model`` is not a :class:`torch.jit.ScriptModule` nor a
   (...)
    501             All errors are subclasses of :class:`errors.OnnxExporterError`.
    502     """
--> 504     _export(
    505         model,
    506         args,
    507         f,
    508         export_params,
    509         verbose,
    510         training,
    511         input_names,
    512         output_names,
    513         operator_export_type=operator_export_type,
    514         opset_version=opset_version,
    515         do_constant_folding=do_constant_folding,
    516         dynamic_axes=dynamic_axes,
    517         keep_initializers_as_inputs=keep_initializers_as_inputs,
    518         custom_opsets=custom_opsets,
    519         export_modules_as_functions=export_modules_as_functions,
    520     )

File ~/miniconda3/envs/inference/lib/python3.8/site-packages/torch/onnx/utils.py:1529, in _export(model, args, f, export_params, verbose, training, input_names, output_names, operator_export_type, export_type, opset_version, do_constant_folding, dynamic_axes, keep_initializers_as_inputs, fixed_batch_size, custom_opsets, add_node_names, onnx_shape_inference, export_modules_as_functions)
   1526     dynamic_axes = {}
   1527 _validate_dynamic_axes(dynamic_axes, model, input_names, output_names)
-> 1529 graph, params_dict, torch_out = _model_to_graph(
   1530     model,
   1531     args,
   1532     verbose,
   1533     input_names,
   1534     output_names,
   1535     operator_export_type,
   1536     val_do_constant_folding,
   1537     fixed_batch_size=fixed_batch_size,
   1538     training=training,
   1539     dynamic_axes=dynamic_axes,
   1540 )
   1542 # TODO: Don't allocate a in-memory string for the protobuf
   1543 defer_weight_export = (
   1544     export_type is not _exporter_states.ExportTypes.PROTOBUF_FILE
   1545 )

File ~/miniconda3/envs/inference/lib/python3.8/site-packages/torch/onnx/utils.py:1115, in _model_to_graph(model, args, verbose, input_names, output_names, operator_export_type, do_constant_folding, _disable_torch_constant_prop, fixed_batch_size, training, dynamic_axes)
   1112 params_dict = _get_named_param_dict(graph, params)
   1114 try:
-> 1115     graph = _optimize_graph(
   1116         graph,
   1117         operator_export_type,
   1118         _disable_torch_constant_prop=_disable_torch_constant_prop,
   1119         fixed_batch_size=fixed_batch_size,
   1120         params_dict=params_dict,
   1121         dynamic_axes=dynamic_axes,
   1122         input_names=input_names,
   1123         module=module,
   1124     )
   1125 except Exception as e:
   1126     torch.onnx.log("Torch IR graph at exception: ", graph)

File ~/miniconda3/envs/inference/lib/python3.8/site-packages/torch/onnx/utils.py:663, in _optimize_graph(graph, operator_export_type, _disable_torch_constant_prop, fixed_batch_size, params_dict, dynamic_axes, input_names, module)
    660     _C._jit_pass_onnx_set_dynamic_input_shape(graph, dynamic_axes, input_names)
    661 _C._jit_pass_onnx_lint(graph)
--> 663 graph = _C._jit_pass_onnx(graph, operator_export_type)
    664 _C._jit_pass_onnx_lint(graph)
    665 _C._jit_pass_lint(graph)

File ~/miniconda3/envs/inference/lib/python3.8/site-packages/torch/onnx/utils.py:1899, in _run_symbolic_function(graph, block, node, inputs, env, operator_export_type)
   1894     if symbolic_fn is not None:
...
--> 100     raise ValueError("Scalar type name cannot be None")
    101 if valid_scalar_name(name):
    102     return _SCALAR_NAME_TO_TYPE[name]  # type: ignore[index]

ValueError: Scalar type name cannot be None
```
